### PR TITLE
Add a flag to turn off all LLVM fast math options.

### PR DESCRIPTION
**Pull request by zvookin**
This is intended to be used to see how much of a difference those options make in both
performance and accuracy. Theoretically StrictFP can be used to generate
more stable results to test against as well.

Any suggestions on tests?

### CODE REVIEW

File: src/CodeGen_Internal.cpp [hunk 1]
```diff
@@ -385,18 +385,20 @@ bool get_md_string(llvm::Metadata *value, std::string &result) {
...
 
-void get_target_options(const llvm::Module &module, llvm::TargetOptions &options, std::string &mcpu, std::string &mattrs) {
+void get_target_options(const llvm::Module &module, llvm::TargetOptions &options,
+                        std::string &mcpu, std::string &mattrs, const Halide::Target &target) {
     bool use_soft_float_abi = false;
     get_md_bool(module.getModuleFlag("halide_use_soft_float_abi"), use_soft_float_abi);
     get_md_string(module.getModuleFlag("halide_mcpu"), mcpu);
     get_md_string(module.getModuleFlag("halide_mattrs"), mattrs);
 
+    bool strict_fp = target.has_feature(Target::StrictFP);
```
Comment (dsharletg) [inline src/CodeGen_Internal.cpp]: I think this new option should be also embedded in the module as metadata (halide_strict_fp or something) like the other options above (particularly halide_use_soft_float_abi which is also a bool). That way, you don't need to plumb around the target (which is most of the changes in this PR :), the llvm::Module is self contained.

Comment (zvookin) [reply src/CodeGen_Internal.cpp]: @abadams Any opinion on this?

Comment (abadams) [reply src/CodeGen_Internal.cpp]: sgtm. It is nice to have a self-contained llvm::Module - i.e. one that tells us what TargetOptions to use for it. The consistency argument is also a strong one

**Comment by steven-johnson** [comment]
https://github.com/halide/Halide/issues/2498

**Comment by steven-johnson** [comment]
LGTM. Would be nice to follow up with some benchmark comparisons.

File: src/IROperator.h [hunk 2]
```diff
@@ -1909,6 +1908,28 @@ Expr saturating_cast(Expr e) {
...
+    Type t = e.type();
+    return Internal::Call::make(t, Internal::Call::no_fp_simplify,
+                                {std::move(e)}, Internal::Call::Intrinsic);
+}
+
+/** Make best effort attempt to preserve IEEE floating-point semantics
+ * in evaluating an expression. Equivalent to no_fp_simplify plus
+ * turning on StrictFP within a certain scope. May not be implemented
+ * for all backends. (E.g. it is difficult to do this for C++ code
+ * generation as it depends on the compielr flags used to compilethe
```
Comment (steven-johnson) [inline src/IROperator.h]: nit: typos

File: src/CodeGen_Internal.h [hunk 3]
```diff
@@ -67,13 +67,14 @@ Expr lower_euclidean_mod(Expr a, Expr b);
...
-void get_target_options(const llvm::Module &module, llvm::TargetOptions &options, std::string &mcpu, std::string &mattrs);
+void get_target_options(const llvm::Module &module, llvm::TargetOptions &options,
+                        std::string &mcpu, std::string &mattrs);
 
 /** Given two llvm::Modules, clone target options from one to the other */
 void clone_target_options(const llvm::Module &from, llvm::Module &to);
 
 /** Given an llvm::Module, get or create an llvm:TargetMachine */
-std::unique_ptr<llvm::TargetMachine> make_target_machine(const llvm::Module &module);
+ std::unique_ptr<llvm::TargetMachine> make_target_machine(const llvm::Module &module);
```
Comment (dsharletg) [inline src/CodeGen_Internal.h]: Remove leading space. In general, there are a lot of changes in this PR that are just reformatting left over from the target plumbing that would be nice (but not necessary) to revert.

File: src/IROperator.h [hunk 4]
```diff
@@ -1909,6 +1908,28 @@ Expr saturating_cast(Expr e) {
...
 EXPORT Expr saturating_cast(Type t, Expr e);
 
+/** Prevents the simplifier from changing floating-point operations
+ * within the argument expression while returning the argument as its result.
+ * Can be used to ensure stricter floating-point semantics but does not by itself
+ * affect lower-level code generation. Mostly provided for debugging. */
+inline Expr no_fp_simplify(Expr e) {
+    Type t = e.type();
+    return Internal::Call::make(t, Internal::Call::no_fp_simplify,
+                                {std::move(e)}, Internal::Call::Intrinsic);
```
Comment (dsharletg) [inline src/IROperator.h]: PureIntrinsic? Not actually sure about that... This is 'meta-side-effecty' but I don't think it's side effecting in the way that violates the meaning of pure that is used to determine if simplifications/optimizations are valid.

File: src/IROperator.h [hunk 5]
```diff
@@ -1909,6 +1908,28 @@ Expr saturating_cast(Expr e) {
...
+/** Make best effort attempt to preserve IEEE floating-point semantics
+ * in evaluating an expression. Equivalent to no_fp_simplify plus
+ * turning on StrictFP within a certain scope. May not be implemented
+ * for all backends. (E.g. it is difficult to do this for C++ code
+ * generation as it depends on the compielr flags used to compilethe
+ * generated code. */
+inline Expr strict_fp(Expr e) {
+    Type t = e.type();
+    return Internal::Call::make(t, Internal::Call::strict_fp,
+                                {std::move(e)}, Internal::Call::Intrinsic);
```
Comment (dsharletg) [inline src/IROperator.h]: PureIntrinsic (similar comment as above).

**Comment by abadams** [comment]
I don't see anything here that prevents CSE and LICM from lifting out the bodies of the dont_simplify intrinsics. They'll get partially pulled out and then simplified.

**Comment by zvookin** [comment]
I'm working on CSE and LICM now. Hoping to have them hoist the scope wrappers so they still do their job, but preserve semantics. CSE will have to be careful to not confuse strict and non-strict values, but that shouldn't be too bad.

**Comment by zvookin** [comment]
Anyone else have an opinion on "StrictFP" vs. "StrictFloat"?

File: src/Simplify.cpp [hunk 6]
```diff
@@ -1746,6 +1783,15 @@ class Simplify : public IRMutator2 {
     Expr visit(const Div *op) override {
         Expr a = mutate(op->a);
         Expr b = mutate(op->b);
+
+        if (no_fp_simplify && op->type.is_float()) {
```
Comment (abadams) [inline src/Simplify.cpp]: Might be simpler in all these cases to put:

if (no_fp_simplify && op->type.is_float()) {
  return IRMutator2::mutate(op);
}

above the lines that mutate a and b

**Comment by shoaibkamil** [comment]
Somewhat tangentially, we should probably also look at controlling GPU fastmath settings (e.g. in Metal, we always turn off fastmath).

**Comment by zvookin** [comment]
Yes, I do intend to hook this up to the GPU controls, in as much as they exist, for strictness. I'm not sure it will happen in the initial PR however.

**Comment by zvookin** [comment]
Reopening this PR here https://github.com/halide/Halide/pull/2668 due to git issues.

**Pull request closed without merging**